File: APP/APPSDK/moda-digitalwallet-holder-iosapp-token-sdk/moda-digitalwallet-holder-sdk/TokenSDKManager.swift
import Foundation
import Flutter

public class TokenSDKManager {
    
    private let methodChannel: FlutterMethodChannel
    private var modaDidKey: ModaDidKey?
    
    public init(messenager: FlutterEngine) {

        methodChannel = FlutterMethodChannel(name: "com.flutter_method_channel/gov.moda.dw", binaryMessenger: messenager.binaryMessenger)
        setMethodCallHandler()
    }
    
    func setMethodCallHandler() {
        methodChannel.setMethodCallHandler(handle)
    }
    
    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init":
            if let args = call.arguments as? Dictionary<String, Any>,
               let tag = args["tag"] as? String,
               let type = args["type"] as? String,
               let pin = args["pin"] as? FlutterStandardTypedData {
                self.initModaDidKey(tag: tag, type: type, pin: pin.data, result: result)
            }else {
                result(FlutterError(code: "UNAVAILABLE",
                                    message: "init arg faild",
                                    details: nil))
            }
        case "getP256Key":
            self.getP256Key(result: result)
        case "verifyUser":
            if let args = call.arguments as? Dictionary<String, Any>,
               let publicKey = args["publicKey"] as? String {
                self.verifyUser(publicKey: publicKey, result: result)
            }else {
                result(FlutterError(code: "UNAVAILABLE",
                                    message: "verifyUser arg faild",
                                    details: nil))
            }
        case "sign":
            if let args = call.arguments as? Dictionary<String, Any>,
               let header = args["header"] as? String,
               let payload = args["payload"] as? String {
                self.sign(header: header, payload: payload, result: result)
            }else {
                result(FlutterError(code: "UNAVAILABLE",
                                    message: "sign arg faild",
                                    details: nil))
            }
        case "delete":
            self.delete(result: result)
        default:
            PrintPro.print("Method not found")
            result(FlutterMethodNotImplemented)
        }
    }
    
    
    private func initModaDidKey(tag: String, type: String, pin :Data, result: FlutterResult) {
        if let sourceType = SourceType(rawValue: type) {
            modaDidKey = ModaDidKey(keyTag: tag, type: sourceType, pin: pin)
            result(true)
        } else {
            result(FlutterError(code: "INVALID_TYPE", message: "Invalid type: \(type)", details: nil))
        }
    }
    
    private func getP256Key(result: FlutterResult){
        if let (publicKey,generateError) = modaDidKey?.getP256Key() {
            if (generateError != nil) {
                result(FlutterError(code: "UNAVAILABLE",
                                    message: generateError?.errorDescription,
                                    details: nil))
            }else {
                result(publicKey)
            }
        }else {
            result(FlutterError(code: "UNAVAILABLE",
                                message: "getP256Key faild",
                                details: nil))
        }
    }
    private func verifyUser(publicKey: String, result: FlutterResult){
        if let (isValid, verifyError) = modaDidKey?.verifyUser(publicKey:publicKey) {
            if (verifyError != nil) {
                result(FlutterError(code: "UNAVAILABLE",
                                    message: verifyError?.errorDescription,
                                    details: nil))
            }else {
                result(isValid)
            }
        }else {
            result(FlutterError(code: "UNAVAILABLE",
                                message: "verifyUser faild",
                                details: nil))
        }
    }
    private func sign(header: String, payload: String, result: FlutterResult){
        if let (signature, signError) = modaDidKey?.sign(header: header, payload: payload) {
            if (signError != nil) {
                result(FlutterError(code: "UNAVAILABLE",
                                    message: signError?.errorDescription,
                                    details: nil))
            }else {
                result(signature)
            }
        }else {
            result(FlutterError(code: "UNAVAILABLE",
                                message: "sign faild",
                                details: nil))
        }
    }
    private func delete(result: FlutterResult){
        if let (isDeleted, deleteError) = modaDidKey?.delete() {
            if (deleteError != nil) {
                result(FlutterError(code: "UNAVAILABLE",
                                    message: deleteError?.errorDescription,
                                    details: nil))
            }else {
                result(isDeleted)
            }
        }else {
            result(FlutterError(code: "UNAVAILABLE",
                                message: "delete faild",
                                details: nil))
        }
    }
}

File: APP/APPSDK/moda-digitalwallet-holder-androidapp-token-sdk/app/src/main/java/com/moda/pkcsnfckit/TokenSDKManager.kt
package com.moda.pkcsnfckit

import android.content.Context
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class TokenSDKManager(flutterEngine: FlutterEngine, context: Context) {

    private val CHANNEL = "com.flutter_method_channel/gov.moda.dw"
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    init {
        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL).setMethodCallHandler{ call, result ->
            when (call.method) {
                "init" -> {
                    scope.launch {
                        val tag = call.argument<String>("tag").toString()
                        val type = call.argument<String>("type").toString()
                        val pin = call.argument<ByteArray>("pin")
                        if (pin!= null && init(tag, type, pin, context)) {
                            result.success(true)
                        }else {
                            result.error("UNAVAILABLE", "init error.", null)
                        }
                    }
                }
                "getP256Key" -> {
                    scope.launch {
                        val p256Key = getP256Key()

                        if (p256Key != "") {
                            result.success(p256Key)
                        } else {
                            result.error("UNAVAILABLE", "p256Key error.", null)
                        }
                    }
                }
                "verifyUser" -> {
                    val myPublicKey = call.argument<String>("publicKey")
                    scope.launch {
                        if (myPublicKey != null) {
                            result.success(verifyUser(myPublicKey))
                        }else {
                            result.error("UNAVAILABLE", "Public key not available.", null)
                        }
                    }
                }
                "sign" -> {
                    val header = call.argument<String>("header")
                    val payload = call.argument<String>("payload")
                    scope.launch {
                        if (header != null && payload != null) {
                            result.success(sign(header, payload))
                        }else {
                            result.error("UNAVAILABLE", "Public key not available.", null)
                        }
                    }
                }
                "delete" -> {
                    scope.launch {
                        if (delete()) {
                            result.success(true)
                        } else {
                            result.error("UNAVAILABLE", "delete error.", null)
                        }
                    }
                }
                "cancel" -> {
                    scope.launch {
                        cancel()
                        result.success(true)
                    }
                }
                else -> result.notImplemented()
            }
        }
    }

    private lateinit var modaDidKey: ModaDidKey

    private fun init(tag: String, type: String, pin: ByteArray, context: Context): Boolean{
        val pinString = String(pin)
        modaDidKey = when (type) {
            "platform" -> ModaDidKey(tag, ModaDidKey.SourceType.PLATFORM, pinString, context)
            else -> return false
        }
        return true
    }

    private suspend fun getP256Key(): String{
        val publicKey = modaDidKey.getP256Key()
        return publicKey.toString()
    }

    private suspend fun verifyUser(publicKey: String): Boolean {
        return modaDidKey.verifyUser(publicKey)
    }

    private suspend fun sign(header: String, payload: String): String{
        val signData = modaDidKey.sign(header,payload)
        return signData.toString()
    }

    private suspend fun delete(): Boolean{
        return modaDidKey.delete()
    }

    private suspend fun cancel() {
        modaDidKey.cancel()
    }
}